apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabTheorycraft/index.tsx

```diff
@@ -286,689 +201,454 @@ export default function TabTheorycraft() {
+            failed,
+            skipped,
+            finishTime: performance.now(),
+          }))
+
+          if (!apply) {
+            console.log({
+              maxBuffer,
+              distributed,
+              tested,
+              failed,
+              skipped,
+            })
+            break
           }
-        />
-      </CardLight>
-    </Stack>
-  )
-}
-function ArtifactSetEditor({
-  setKey,
-  value,
-  setValue,
-  deleteValue,
-  remaining,
-}: {
-  setKey: ArtifactSetKey
-  value: 1 | 2 | 4
-  setValue: (v: 1 | 2 | 4) => void
-  deleteValue: () => void
-  remaining: number
-}) {
-  const artifactSheet = getArtSheet(setKey)
 
-  /* Assumes that all conditionals are from 4-Set. needs to change if there are 2-Set conditionals */
-  const set4CondNums = useMemo(() => {
-    if (value < 4) return []
-    return Object.keys(artifactSheet.setEffects).filter((setNumKey) =>
-      artifactSheet.setEffects[setNumKey]?.document.some(
-        (doc) => 'states' in doc
+          setCharTC((charTC) => {
+            charTC.artifact.substats.stats = objMap(
+              charTC.artifact.substats.stats,
+              (v, k) => v + toPercent(maxBuffer![k] ?? 0, k)
+            )
+            charTC.optimization.distributedSubstats =
+              distributedSubstats - distributed
+            return charTC
+          })
+
+          break
+        }
+      }
+    }
+  }
+
+  const kqms = useCallback(() => {
+    setCharTC((charTC) => {
+      charTC.artifact.substats.type = 'mid'
+
+      const {
+        artifact: {
+          slots,
+          substats: { stats, type, rarity },
+        },
+      } = charTC
+      charTC.optimization.distributedSubstats =
+        20 - (rarity === 5 ? 0 : rarity === 4 ? 10 : 15)
+      charTC.artifact.substats.stats = objMap(stats, (val, statKey) => {
+        const substatValue = getSubstatValue(statKey, rarity, type)
+        return substatValue * 2
+      })
+      charTC.optimization.maxSubstats = objMap(
+        charTC.optimization.maxSubstats,
+        (_val, statKey) => {
+          const rollsPerSlot = 2
+          const diffSlot = 5
+          const sameSlot = Object.entries(slots).filter(
+            ([_slotKey, { statKey: mainStatKey }]) => mainStatKey === statKey
+          ).length
+          // +2 for the inital fixed rolls
+          return 2 + (diffSlot - sameSlot) * rollsPerSlot
+        }
       )
-    )
-  }, [artifactSheet, value])
+    })
+  }, [setCharTC])
 
   return (
-    <CardLight>
-      <Box display="flex">
-        <ArtifactSetTooltip artifactSheet={artifactSheet} numInSet={value}>
-          <Box flexGrow={1} p={1} display="flex" gap={1} alignItems="center">
-            <ImgIcon size={2} src={artifactDefIcon(setKey)} />
-            <Box>{artifactSheet.setName}</Box>
-            <Info />
+    <CharTCContext.Provider value={valueCharTCContext}>
+      <Stack spacing={1}>
+        <CardLight>
+          <Box sx={{ display: 'flex', gap: 1, p: 1 }}>
```

Comment:
I think you want cardcontent?

Reply:
nope, cardcontent has `2` padding, this needed `1`